Implement ConcurrentHashMap.newKeySet and use it in VersionChecks

exoego opened the pull request:
closes #3882

exoego commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -203,34 +203,62 @@ object ConcurrentHashMap {
       extends Set[K] with Serializable {
 
-    def contains(o: Any): Boolean = inner.contains(o)
+    def getMappedValue(): V = defaultValue
```
getMappedValue is a public member of KeySetView according to https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.KeySetView.html#getMappedValue--

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -194,6 +194,30 @@ class ConcurrentHashMapTest extends MapTest {
     }
   }
+
+  @Test def testNewKeySet(): Unit = {
```
Instead of this, use the tests in `SetTest`?

exoego replied:
Added https://github.com/scala-js/scala-js/pull/4367/commits/7c6cd273e35808874878876d8828ae05a0064bf7

I will remove this tests after the `SetTest`-based tests passed.

exoego replied:
Removed

exoego commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -203,34 +203,62 @@ object ConcurrentHashMap {
+      true
+    } else {
+      false
+    }
```
Done in https://github.com/scala-js/scala-js/pull/4367/commits/6a1dd4b2dcd8a002acfa72c3e479fb37baf8e149

exoego commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -203,34 +203,62 @@ object ConcurrentHashMap {
+    } else {
+      innerMap.put(e, defaultValue)
+      true
+    }
```
Done in https://github.com/scala-js/scala-js/pull/4367/commits/aeadef3145f2cceebf73cf2d9724d8c4a67345c4

exoego commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -203,34 +203,62 @@ object ConcurrentHashMap {
    * to the inner key set.
    */
-  class KeySetView[K, V] private[ConcurrentHashMap] (inner: Set[K])
+  class KeySetView[K, V] private[ConcurrentHashMap] (innerMap: ConcurrentHashMap[K, V], defaultValue: V)
```
I think constructor requires 2nd parameter, because
* The instance method `ConcurrentHashMap#keySet()` passes `null.asInstanceOf[V]` to it
* The static method `newKeySet[K](): KeySetView[K, Boolean]` passes `true` to it

gzm0 replied:
Yes, that is correct. But why are you passing a `ConcurrentHashMap` as the first parameter and not a `InnerHashMap`? This seems unnecessary indirection.

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -194,6 +194,38 @@ class ConcurrentHashMapTest extends MapTest {
+    assertThrows(classOf[UnsupportedOperationException], keySet.retainAll(null))
+    assertThrows(classOf[UnsupportedOperationException], keySet.removeAll(null))
+    assertThrows(classOf[UnsupportedOperationException], keySet.add(null))
+    assertThrows(classOf[UnsupportedOperationException], keySet.addAll(null))
```
Are you sure these test cases are correct? [`ConcurrentHashMap.containsKey`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#containsKey-java.lang.Object-) should throw an NPE in this case.

exoego commented:
Hmm, `sbt testSuite2_13/test` failed due to missing `newKeySet`, which is added in this PR.

```
[error] C:\projects\scala-js\test-suite\shared\src\test\scala\org\scalajs\testsuite\javalib\util\ConcurrentHashSetTest.scala:60:41: type newKeySet is not a member of object java.util.concurrent.ConcurrentHashMap
[error]     new ju.concurrent.ConcurrentHashMap.newKeySet[E]()
[error]                                         ^
[error] one error found
```

exoego commented:
It is werid that `ConcurrentHashMapKeySetViewTest` is failing...

```
val hs = factory.empty[String]

assertEquals(0, hs.size) // pass
assertTrue(hs.add("ONE")) // fail with AssertionError: null
```

I do not figure out where `AssertionError` is coming from.

gzm0 commented:
`AssertionError: null` may mean that the assert itself is failing (i.e. `add` returns `false` instead of `true`). The `null` simply means the `message` field is `null`. Which happens if `assertTrue` doesn't have a second `msg` parameter.

gzm0 commented:
Re: `keySet()` without parameters: It seems this is specified in `Map` and quite well tested in `MapTest`. However, that `add` and `addAll` throw an `UnsupportedOperationException` is not tested there, but it probably should. So IMO the best course of action is to add these tests there (and fix offending implementations if any).

exoego commented:
> So IMO the best course of action is to add these tests there (and fix offending implementations if any).

Added tests and fixed implementations.

I will squash commits into single comment when PR get approved.

gzm0 commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -199,38 +199,65 @@ object ConcurrentHashMap {
+        if (!innerMap.containsKey(nextKey)) {
+          innerMap.put(nextKey, defaultValue)
+          addedAny = true
+        }
```
`addedAny ||= innerMap.putIfAbsent(iter.next(), defaultValue)`?

exoego replied:
Hmm, I tried the below 

```scala
      while (iter.hasNext()) {
        val nextKey = iter.next()
        addedAny ||= innerMap.putIfAbsent(nextKey, defaultValue) == null
      }
```
but this made some tests failing.

gzm0 replied:
Potentially the same issue than for `add`.

exoego replied:
The below tests are failing, even after fixed `NullRejectingHashMap.putIfAbsent`.

```
[error] Test org.scalajs.testsuite.javalib.util.ConcurrentHashMapKeySetViewTest.simpleSum failed: java.lang.AssertionError: expected:<268> but was:<42>, took 0.001999872 sec
...
[error] Test org.scalajs.testsuite.javalib.util.ConcurrentHashMapKeySetViewTest.toStringCollectionCustomClass failed: java.lang.AssertionError: result '[Custom(A,1)]' not in expected set 'HashSet([Custom(b,2), Custom(A,1), Custom(C,3)], [Custom(b,2), Custom(C,3), Custom(A,1)], [Custom(C,3), Custom(b,2), Custom(A,1)], [Custom(C,3), Custom(A,1), Custom(b,2)], [Custom(A,1), Custom(b,2), Custom(C,3)], [Custom(A,1), Custom(C,3), Custom(b,2)])', took 0.004 sec
...
[error] Test org.scalajs.testsuite.javalib.util.ConcurrentHashMapKeySetViewTest.toStringCollectionDoubleHasCommaSpace failed: java.lang.AssertionError: result '[88.42]' not in expected set 'HashSet([60.173, 88.42, -23.36], [88.42, -23.36, 60.173], [-23.36, 60.173, 88.42], [60.173, -23.36, 88.42], [88.42, 60.173, -23.36], [-23.36, 88.42, 60.173])', took 0.001000192 sec
...
[error] Test org.scalajs.testsuite.javalib.util.ConcurrentHashMapKeySetViewTest.removeIf failed: java.lang.AssertionError: expected:<8> but was:<1>, took 0 sec
...
[error] Test org.scalajs.testsuite.javalib.util.ConcurrentHashMapKeySetViewTest.iterator failed: java.lang.AssertionError: expected:<3> but was:<1>
```

gzm0 replied:
Why are they failing? Also, I just realize that using `add` might even be a better option. Like here: https://github.com/scala-js/scala-js/blob/5cc821bd51ba49d7ee33e5a2dffa44636864da41/javalib/src/main/scala/java/util/HashSet.scala#L57-L63

exoego replied:
Ah, that one using `add` worked well.

gzm0 replied:
Ah, that's my bad :) The problem was short-circuiting behavior of `||`.

gzm0 commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -199,38 +199,65 @@ object ConcurrentHashMap {
+        if (!innerMap.containsKey(nextKey)) {
+          innerMap.put(nextKey, defaultValue)
+          addedAny = true
+        }
```
`addedAny ||= innerMap.putIfAbsent(e, defaultValue)`?

gzm0 commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -199,38 +199,65 @@ object ConcurrentHashMap {
 
-    def addAll(c: Collection[_ <: K]): Boolean = inner.addAll(c)
+    def addAll(c: Collection[_ <: K]): Boolean = {
+      if (c == null || defaultValue == null) {
```
Same here about the `null` check. Just leave it to `c.iterator()`.

gzm0 commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -199,38 +199,65 @@ object ConcurrentHashMap {
 
-    def add(e: K): Boolean = inner.add(e)
+    def add(e: K): Boolean = {
+      if (e == null || defaultValue == null) {
```
I'm pretty positive the `null` check here is wrong. If `c` is `null`, this should throw an `NPE`, not an `UnsupportedOperationException`.

I'd just leave the `null` check to `putIfAbsent`.

exoego replied:
Hmm, but removing `e == null` fails on `SetTest#contains`.
It seems that `putIfAbsent` on InnerHashMap throw nothing on `null` key.

gzm0 replied:
Hah, indeed. It seems like `ConcurrentHashMap` explicitly fixes this behavior. Was this an oversight in f7be410503f87b402a86c92945dba37455f06853? @sjrd ? If yes, maybe this should be moved to `NullRejectingHashMap`.

sjrd replied:
Ah yes, well spotted. Indeed, introducing the optimized implementation of `putIfAbsent` in `HashMap` bypassed the null checks that `NullRejectingHashMap` performs. So indeed, we should override `putIfAbsent` in `NullRejectingHashMap` to perform the null checks before calling `super.putIfAbsent` (similarly to `put`).

exoego replied:
Fixed in https://github.com/scala-js/scala-js/pull/4367/commits/16687e688329b3ed7c612f33e030517278d10a2a

gzm0 replied:
Hum, did this line also change? `e == null` here is definitely wrong. If `e == null`, this should throw an NPE (but `innerMap.putIfAbsent` will already do so).

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -194,6 +194,17 @@ class ConcurrentHashMapTest extends MapTest {
+    expectThrows(classOf[UnsupportedOperationException], set.addAll(null))
+    expectThrows(classOf[UnsupportedOperationException], set.addAll(ju.Arrays.asList(null)))
+    expectThrows(classOf[UnsupportedOperationException], set.addAll(ju.Arrays.asList("ONE")))
+  }
```
This should go to `MapTest`.

exoego replied:
Moved to `MapTest` https://github.com/scala-js/scala-js/pull/4367/commits/d636996881b26218886a3f03952105e710b408a9

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -194,6 +194,17 @@ class ConcurrentHashMapTest extends MapTest {
         assertTrue(s"missing value '$value'", encounteredValues.contains(value))
     }
   }
+
```
There should be at least one test case checking the `add` functionality of `ConcurentHashMap.KeySetView`. Notably the following behavior should be tested:

- Adding a new key adds the default value.
- Adding an existing key does *not* alter the value (iff this is the JVM behavior, the doc is IMO vague about it).

exoego replied:
Hmm, I do not understand.

`KeySetView[K, V]` have only `getMappedValue` to return `V`.
And `KeySetView` made from `ConcurrentHashMap` do not support `add/addAll`.
So, I do not think it is possible to check

* the default value is added when a new key is added
* the value is not altered when an exsting key is added

gzm0 replied:
Oh, that is supported if the overload specific to `ConcurrentHashMap` of `keySet` is used: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#keySet-V-

exoego replied:
It seems that any addition to the `keySet` always throws Exception.
https://github.com/scala-js/scala-js/pull/4367/commits/d636996881b26218886a3f03952105e710b408a9

exoego replied:
Ah, do you mean using private constructor `KeySetView(map, defaultValue)` ?

exoego replied:
Ah, I see.

exoego commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -199,38 +199,65 @@ object ConcurrentHashMap {
+
+    def getMappedValue(): V = defaultValue
+
+    override def toString: String = innerMap.keySet().toString
```
Filed in https://github.com/scala-js/scala-js/issues/4395
I will open a separate small PR for this.

exoego replied:
Fixed

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/MapTest.scala:
```diff
@@ -1419,6 +1419,23 @@ trait MapTest {
+    if (factory.allowsNullCollectionsAdding) {
+      expectThrows(classOf[UnsupportedOperationException], set.addAll(null))
+    } else {
+      // NPE on JVM, JSE on JS
```
If the JVM exposes different behavior when passing `null` collection, we should probably just drop this specific test (it's arguably a degenerate case).

@sjrd for a second opinion.

sjrd replied:
Yes, this is full-blown Undefined Behavior, because the *collection* itself is null, not its elements. This test should be dropped (also the "then" branch, and therefore the `allowsNullCollectionsAdding` factory flag).

exoego replied:
Removed assertion.
But it seems that `coll.addAll(nulll)` should throw NPE according to JavaDoc.
So filed https://github.com/scala-js/scala-js/issues/4399

gzm0 replied:
This isn't addressed. The entire `if` and `allowsNullCollectionsAdding` should be removed.

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/MapTest.scala:
```diff
@@ -1419,6 +1419,23 @@ trait MapTest {
+
+  @Test def additionToKeySet(): Unit = {
+    val chm = factory.empty[String, String]
+    val set = chm.keySet()
```
Simply `val set = factory.empty[String, String].keySet()`? (`chm` is unused later).

exoego replied:
Fixed in https://github.com/scala-js/scala-js/pull/4367/commits/4c21243134bca29258c37d1618657d561e6941cd

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -194,6 +194,34 @@ class ConcurrentHashMapTest extends MapTest {
+    assertEquals("Adding a new key adds the default value.", 0, map.get("ONE"))
+
+    assertFalse(keySet.add("ONE"))
+    assertEquals("Adding an existing key do not alter the value", 0, map.get("ONE"))
```
The message here is misleading: Even *if* the value was altered, it would still be `0`. Consider simply dropping this assert.

exoego replied:
Removed in https://github.com/scala-js/scala-js/pull/4367/commits/1940ba1fc571b40be360be116f980a7f99d9a62b

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -194,6 +194,34 @@ class ConcurrentHashMapTest extends MapTest {
+    assertEquals("Adding an existing key do not alter the value", 0, map.get("ONE"))
+
+    assertEquals(0, map.put("ONE", 42))
+    assertEquals("Putting alter the value", 42, map.get("ONE"))
```
Nit: alter*s*

exoego replied:
Fixed in https://github.com/scala-js/scala-js/pull/4367/commits/873d05ed0a4fd4c60f549c666da8f20abf3515ef

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -194,6 +194,34 @@ class ConcurrentHashMapTest extends MapTest {
+    assertEquals("Putting alter the value", 42, map.get("ONE"))
+
+    assertFalse(keySet.add("ONE"))
+    assertEquals("Adding an existing key do not alter the value", 42, map.get("ONE"))
```
Nit: ... key do*es* not ...

exoego replied:
Fixed in https://github.com/scala-js/scala-js/pull/4367/commits/873d05ed0a4fd4c60f549c666da8f20abf3515ef

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -194,6 +194,34 @@ class ConcurrentHashMapTest extends MapTest {
+    assertEquals("[a]", keySet.toString)
+
+    map.put("b", 1)
+    assertEquals("[a, b]", keySet.toString)
```
This test is likely overly restrictive: IIUC there is not guarantee of any particular key ordering when iterating over a `KeySetView`. So either `"[a, b]"` or `"[b, a]"` would be correct in this case.

exoego replied:
Fixed in https://github.com/scala-js/scala-js/pull/4367/commits/afbb9c57f285c8d60f66625178d97a4df1071d69

gzm0 commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -62,7 +62,10 @@ class ConcurrentHashMap[K, V] private (initialCapacity: Int, loadFactor: Float)
 
   override def keySet(): ConcurrentHashMap.KeySetView[K, V] =
-    new ConcurrentHashMap.KeySetView[K, V](inner.keySet())
+    keySet(null.asInstanceOf[V])
```
Actually, this usage seems to be not allowed: Doc says that `keySet(V)` throws NPE if `mappedValue` is `null`. We should probably honor that specification to avoid an accidental `null` to be interpreted as a sentinel.

gzm0 commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -62,7 +62,10 @@ class ConcurrentHashMap[K, V] private (initialCapacity: Int, loadFactor: Float)
 
   override def keySet(): ConcurrentHashMap.KeySetView[K, V] =
-    new ConcurrentHashMap.KeySetView[K, V](inner.keySet())
+    keySet(null.asInstanceOf[V])
```
@sjrd is usage of `null` like this problematic if `V` is a primitive type?

sjrd replied:
No, because it won't escape the generic environment, so it will always remain in places that erase to `Object`.

If it *did* escape into a monomorphic context, it would unbox to the zero of the target type.

exoego replied:
Hmm,

```scala
val map = new java.util.concurrent.ConcurrentHashMap[String,String]()
val set = map.keySet(null)
```
does not throw NPE.

gzm0 replied:
That clearly contradicts the documentation: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#keySet-V-

On JDK 11, it does throw for me. If if doesn't in earlier JVM versions, we should add an `assumeTrue` for the JDK.

exoego replied:
Addressed in  https://github.com/scala-js/scala-js/pull/4367/commits/1e8f329d2492c2a496cbab2d68e5a0f3d67714c9

gzm0 commented:
What is the status on this? @exoego is there anything you need from our end or are you unblocked?

exoego commented:
I think that all mentioned issues were addressed.

exoego commented on javalib/src/main/scala/java/util/NullRejectingHashMap.scala:
```diff
@@ -57,6 +57,12 @@ private[util] class NullRejectingHashMap[K, V](
 
+  override def putIfAbsent(key: K, value: V): V = {
+    if (key == null)
+      throw new NullPointerException()
```
Added assertions https://github.com/scala-js/scala-js/pull/4367/commits/fa1678d61c4b6f19dfee95000e95c28b9ff135b0
It seems that `putIfAbsent` already throws when `null` value.

gzm0 replied:
Well, in `ConcurrentHashMap` it does, because you it still has the old implementation rather than forwarding to this one.

gzm0 replied:
In fact, now that I look at it, the implementation should be moved for `remove` and `replace` as well.

exoego replied:
Added assertions in https://github.com/scala-js/scala-js/pull/4367/commits/d1aa9be1f7713c983a60c2baebaba10444c4d27d
It seems that `remove` and `replace` already throws  NPE, so overriding is seems not required.

gzm0 replied:
They only throw an NPE because that's what the implementation in ConcurrentHashMap does. You need to change the implementations in ConcurrentHashMap to simply forward to NullRejectingHashMap. (that applies to `putIfAbsent`, `remove` and `replace`). Then you'll see that `remove` and `replace` also need updating.

exoego replied:
Moved implementation to NullRejectingHashMap in https://github.com/scala-js/scala-js/pull/4367/commits/811bd513c9b9a461ddb88969460173663d981de3

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/MapTest.scala:
```diff
@@ -1200,6 +1200,12 @@ trait MapTest {
       assertNull(mp.putIfAbsent("nullable", "non null"))
       assertEquals("non null", mp.get("nullable"))
+    } else {
+      assertThrows(classOf[NullPointerException], mp.putIfAbsent("abc", null))
```
Also add a case where the key isn't present yet?

exoego commented on javalib/src/main/scala/java/util/NullRejectingHashMap.scala:
```diff
@@ -33,6 +35,10 @@ private[util] class NullRejectingHashMap[K, V](
+  override def compute(key: K, remappingFunction: BiFunction[_ >: K, _ >: V, _ <: V]): V = {
+    super.compute(key, remappingFunction)
+  }
+
```
Oops.
Removed in https://github.com/scala-js/scala-js/pull/4367/commits/9191b6c746d6b84f742ddc430e0632fae75e7db2

exoego commented on javalib/src/main/scala/java/util/NullRejectingHashMap.scala:
```diff
@@ -12,6 +12,8 @@
 package java.util
 
+import java.util.function.BiFunction
+
```
Removed in https://github.com/scala-js/scala-js/pull/4367/commits/1c26ee4aa9149ac1133d51228219ab64944f872f

exoego commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/MapTest.scala:
```diff
@@ -1217,12 +1227,43 @@ trait MapTest {
       assertTrue(mp.remove("nullable", null))
       assertFalse(mp.containsKey("nullable"))
+    } else {
+      assertThrows(classOf[NullPointerException], mp.remove(null, "old value"))
```
Removed assertions for NPE https://github.com/scala-js/scala-js/pull/4367/commits/b42213dd1509be1cc154a5738e11c6c4ae4078ec
Both [JavaDoc JDK 8](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#remove-java.lang.Object-java.lang.Object-) and [JavaDoc JDK 11](
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ConcurrentHashMap.html#remove(java.lang.Object,java.lang.Object) does not mention `chm#remove(key, null)` should throw NPE.

Instead, it seem that `chm#remove(key, null)` should return `false`.
https://github.com/scala-js/scala-js/pull/4367/commits/01ac6d2bf060895b8fd5be3645d60989439c365b

exoego commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -194,6 +196,42 @@ class ConcurrentHashMapTest extends MapTest {
     }
   }
+
+  @Test def keySetWithMappedValue(): Unit = {
```
FIxed so in https://github.com/scala-js/scala-js/pull/4367/commits/78fd5a98f32ff8ee132accfc1df326fa33139d0d

exoego commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -194,6 +196,42 @@ class ConcurrentHashMapTest extends MapTest {
   }
+
+  @Test def keySetWithMappedValue(): Unit = {
+    assumeTrue("Assumed JS or JDK>8", !executingInJVM || !executingInJVMOnJDK8OrLower)
```
https://github.com/scala-js/scala-js/pull/4367/commits/21ff1ea6db82f7b27b0db10e247a9b13abe24d3b
Removed `assume`, since `keySet(null)` throwing NPE on JDK 8 on my end too.

exoego commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -194,6 +196,42 @@ class ConcurrentHashMapTest extends MapTest {
   }
+
+  @Test def keySetWithMappedValue(): Unit = {
+    assumeTrue("Assumed JS or JDK>8", !executingInJVM || !executingInJVMOnJDK8OrLower)
```
Same as https://github.com/scala-js/scala-js/pull/4367#discussion_r578378572

gzm0 commented on javalib/src/main/scala/java/util/NullRejectingHashMap.scala:
```diff
@@ -80,6 +89,37 @@ private[util] class NullRejectingHashMap[K, V](
+      super.put(key, value)
+    old
+  }
+
```
Some of the `null` handling is inconsistent here: Sometimes it is delegated to other methods (notably `get`) when possible, sometimes not. Have you considered simply moving the implementations that are currently in ConcurrentHashMap.

exoego replied:
CHM now simply delegets to NullRejectingHashMap

gzm0 replied:
Good. But the `null` handling is still inconsistent: Some methods delegate to `get` to check if they `key` is non-null, some don't.

For simplicity: Simply move the implementations from `ConcurrentHashMap`?

exoego replied:
Delegate null check to `(this.)get|` or other method as much s possible. https://github.com/scala-js/scala-js/pull/4367/commits/f12100cd44447c752fc3402e05438eab68e12ed8

gzm0 replied:
Now some of the comments are gone. Please just take the implementations in CHM.

exoego replied:
https://github.com/scala-js/scala-js/pull/4367/commits/fa361b53962fb1b4a686f8a8093a61c04aa42a3b

gzm0 commented on javalib/src/main/scala/java/util/NullRejectingHashMap.scala:
```diff
@@ -80,6 +89,37 @@ private[util] class NullRejectingHashMap[K, V](
   }
 
+  override def remove(key: Any, value: Any): Boolean = {
+    val old = this.get(key) // throws if `key` is null
```
We usually do not qualify `this` if not necessary.

exoego replied:
Removed `this.` qualifier https://github.com/scala-js/scala-js/pull/4367/commits/9e530f3c125e9a1af2d9f12e238a1f5eea3f5f55

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/MapTest.scala:
```diff
@@ -1217,12 +1228,48 @@ trait MapTest {
+      // JavaDoc does not describe that CHM#remove(key, null) should throw NPE, and actually
+      // they do not throw NPE both on JDK 8 & 11. So we do not assert NPE thrown.
+      // However, it should be asserted that mp#(key, null) should not remove.
+      // https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6272521
```
This is not the right comment style. Mutli line comments should use: `/*` (see the coding style guidelines).

I also don't necessarily agree with your interpretation that the spec is ambiguous (or at least this is how I read the text). IMO if it *doesn't* say it throws, it shouldn't. So maybe, the comment can simply be removed.

exoego replied:
Updated https://github.com/scala-js/scala-js/pull/4367/commits/cd4ce0f68f72fcfce8b0bb21100bba50c97f6dbe

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -19,10 +19,12 @@ import java.{util => ju}
 
 import org.scalajs.testsuite.javalib.util.MapTest
 import org.scalajs.testsuite.utils.AssertThrows._
+import org.scalajs.testsuite.utils.Platform._
```
These imports are now unused.

exoego replied:
Deleted https://github.com/scala-js/scala-js/pull/4367/commits/88509a247e7afbf21fff0f12de96c1d4160feac3

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -19,6 +19,7 @@ import java.{util => ju}
 import java.util.concurrent.{ConcurrentHashMap => CHM}
 
 import org.junit.Assert._
+import org.junit.Assume._
```
This import is unused.

exoego replied:
Removed https://github.com/scala-js/scala-js/pull/4367/commits/0eb6a4b9292ebea8e6883759e56119111c994c2f

gzm0 commented:
LGTM! go for squash :)

exoego commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -198,39 +178,58 @@ object ConcurrentHashMap {
+      val iter = c.iterator()
+      var changed = false
+      while (iter.hasNext())
+        changed = innerMap.putIfAbsent(iter.next(), defaultValue) == null || changed
```
@gzm0 
This was `changed = add(iter.next() == null || changed` before squash.
I think null check on `defaultValue` was redundant here, so rewrote.

gzm0 replied:
yep, good find.

gzm0 commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -79,45 +87,17 @@ class ConcurrentHashMap[K, V] private (initialCapacity: Int, loadFactor: Float)
-    old
-  }
+  override def putIfAbsent(key: K, value: V): V =
+    inner.putIfAbsent(key, value)
```
These changes should go into the next commit (all the null check moves should).

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:
```diff
@@ -194,6 +194,40 @@ class ConcurrentHashMapTest extends MapTest {
+    assertThrows(classOf[NullPointerException], map.keySet(null))
+  }
+
+
```
Double blank linke

gzm0 reviewed (changes requested):
Re commit messages:

- Use `Fix #xxx:`, not `Close #xxx:`.
- Message for #4395 should qualify `toString`. (`Fix #4395: ConcurrentHashMap.KeySetView.toString should print elements`)

exoego commented:
@gzm0 I think all issues on previous squash were addressed.

gzm0 commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -198,39 +206,58 @@ object ConcurrentHashMap {
+      innerMap.putIfAbsent(e, defaultValue) == null
+    }
 
     def containsAll(c: Collection[_]): Boolean = inner.containsAll(c)
```
This line should be removed in this commit.

gzm0 commented on javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala:
```diff
@@ -232,7 +232,7 @@ object ConcurrentHashMap {
       innerMap.putIfAbsent(e, defaultValue) == null
     }
 
-    def containsAll(c: Collection[_]): Boolean = inner.containsAll(c)
```
This diff needs to go in the previous commit.

exoego replied:
Ah, sorry. Fixed

sjrd commented:
And we're down under the bar of 40 open issues! 🎉 Thank you @exoego !
